### DISCUSSION TITLE
Add type tests

### DISCUSSION THREAD

**Discussion by stefaanv** [radashi-org/radashi#118] [Ideas]
I haven't always been entirely satisfied with radashi's return types.
Would it be a good idea to add type tests ?
I found basic information on type testing [here](https://vitest.dev/guide/testing-types).
Also needs some changes to `package.json` script section. 

Here's a trivial example for the `alphabetical` function
```ts
import * as _ from 'radashi'
import { test, describe, expectTypeOf } from 'vitest'

type NameAndAge = { name: string; age: number }

describe('alphabetical function type test', () => {
  const list: Array<NameAndAge> = [
    { name: 'Leo', age: 28 },
    { name: 'AJ', age: 25 },
    { name: 'Cynthia', age: 30 }
  ]
  const result = _.alphabetical(list, i => i.name)

  test('returns an array of items with the same type as the original array', () => {
    expectTypeOf(result).toBeArray()
    expectTypeOf(result).items.toMatchTypeOf<NameAndAge>()
    // @ts-expect-error name is a string
    expectTypeOf(result).items.toMatchTypeOf<{ name: boolean }>()
  })
})
```

**Comment by aleclarson** [comment]
We are using this exact Vitest API for type testing in a few places already.
- [select](https://github.com/radashi-org/radashi/blob/main/tests/array/select.test-d.ts)
- [sift](https://github.com/radashi-org/radashi/blob/main/tests/array/sift.test-d.ts)
- [isArray](https://github.com/radashi-org/radashi/blob/main/tests/typed/isArray.test-d.ts)
- [isMap](https://github.com/radashi-org/radashi/blob/main/tests/typed/isMap.test-d.ts)
- …probably some others

I believe we only need them for functions with complex return types (a.k.a. “generics”). I'm totally open to adding more where it makes sense! 👍

**Comment by stefaanv** [comment]
@aleclarson 
Could it be that the ` --typecheck` flag is missing in the `test` script in `package.json` ?
On my computer, the type tests don't run unless I manually add the flag
```json
"scripts": {
  //...
  "test": "vitest run --coverage  --typecheck"
  //...
}
```

**Reply by aleclarson** [reply]
The type tests are checked when you run `pnpm lint`.

I'd like to switch to using Vitest's `--typecheck` flag when the feature is complete. It's supposed to have better debuggability (because of its error formatting), but in my test this morning, it failed to report an issue that was caught by `pnpm lint`. For now, we'll keep using `tsc` directly in `pnpm lint`.
